File: src/lex.c
//
// lex.c - lexical analysis
//

#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

#include "lex.h"
#include "keyw.h"

Lex *lex_make(void);
void lex_readfrom(const char *);
Token *lex_next(void);
static char next(void);
static char peek(void);
static void ignore(void);
static Token *emit(TokenType);
static Token *lex_keyword(void);
static Token *lex_word(void);
static Token *lex_number(void);
static Token *lex_and(void);
static Token *lex_or(void);
static Token *lex_semi(void);
static Token *lex_less(void);
static Token *lex_great(void);
static void lex_space(void);

// ---------------------------------------------------------------------------

static Lex *l;

// lex_make allocates and returns a Lex struct. Its field are just read-only,
// make sure not setting any of its fields.  lex_make has to be called before
// using the rest of public functions listen in "lex.h".
Lex *lex_make(void)
{
    l = malloc(sizeof(Lex));
    l->pos = 0;
    l->stt = 0;
    l->seen[0] = TEOF;
    l->seen[1] = TEOF;
    l->seen[2] = TEOF;
    l->done = true;
    return l;
}

// lex_readfrom sets Lex->buf to point to the provided input by the
// caller and reset Lex->pos, Lex->stt, and Lex->done to its zero values.
void lex_readfrom(const char *input)
{
    l->buf = input;
    l->pos = 0;
    l->stt = 0;
    l->done = false;
}

// next returns the next character in the buf.
static char next(void)
{
    if (!l->buf[l->pos]) {
	l->done = true;
    }

    char c = l->buf[l->pos];
    l->pos++;
    return c;
}

// peek returns but does not consume the next character in the buf.
static char peek(void)
{
    char c = next();
    l->pos--;
    return c;
}

// ignore skips over the pending buf before this point.
static void ignore(void)
{
    l->stt = l->pos;
}

// emit returns a token back to the caller.
static Token *emit(TokenType type)
{
    // Update the three last seen token types.
    l->seen[2] = l->seen[1];
    l->seen[1] = l->seen[0];
    l->seen[0] = type;

    // Prepare the current token.
    Token *t = malloc(sizeof(Token));
    t->type = type;

    int n = l->pos - l->stt;
    t->text = malloc(sizeof(char) * (n + 1));
    t->text[n] = '\0';

    strncpy(t->text, l->buf + l->stt, n);
    l->stt = l->pos;
    return t;
}

// lex_next returns the next token available in buf.
Token *lex_next(void)
{
    for (;;) {
	switch (next()) {

	default:
	    return lex_word();

	case '&':
	    return lex_and();

	case '|':
	    return lex_or();

	case ';':
	    return lex_semi();

	case '<':
	    return lex_less();

	case '>':
	    return lex_great();

	    // These are all of the possible characters a keyword can start with.
	case 'i':
	case 't':
	case 'e':
	case 'f':
	case 'd':
	case 'c':
	case 'u':
	case 'w':
	case '{':
	case '}':
	case '!':
	    return lex_keyword();

	case '0':
	case '1':
	case '2':
	case '3':
	case '4':
	case '5':
	case '6':
	case '7':
	case '8':
	case '9':
	    return lex_number();

	case ' ':
	case '\t':
	case '\v':
	case '\f':
	case '\r':
	case '\n':
	    lex_space();
	    break;

	case '\0':

	    // The null-terminated character has already been reached from the
	    // buf, which means, there is no more buf to read from.
	    return emit(TEOF);
	}
    }
}

// lex_keyword scans:  TIf      TThen    TElse    TElif    TFi   TDo   TDone.
//                     'if'     'then'   'else'   'elif'   'fi'  'do'  'done'
//
//                     TCase    TEsac    TWhile   TUntil   TFor
//                     'case'   'esac'   'while'  'until'  'for'
//
//                     TLBrace  TRBrace  TBang
//                     '{'      '}'      '!'
static Token *lex_keyword(void)
{
    for (;;) {
	switch (peek()) {

	default:

	    // If calling peek() doesn't give a character of the set of the 
	    // character that a keyword can have in it or a space character,
	    // continue processing the token as a TWord.
	    return lex_word();

	    // These are all of the possible characters that 
	    // no-a-single-character keyword can have in it.
	case 'a':
	case 'c':
	case 'd':
	case 'e':
	case 'f':
	case 'h':
	case 'i':
	case 'l':
	case 'n':
	case 'o':
	case 'r':
	case 's':
	case 't':
	case 'u':
	case 'w':
	    next();
	    break;

	case ' ':
	case '\t':
	case '\v':
	case '\f':
	case '\r':
	case '\n':
	case '\0':
	    Token * t = emit(TWord);

	    // At this point, we have a t->text which has only characters 
	    // of the set of characters that a keyword can have in it. We
	    // have to make sure it is a keyword, otherwise, it is a Word.
	    TokenType type = keyw_typeof(t->text);

	    // The type of a 'in' token is a TIn if and only if the third 
	    // last token type is TFor or TCase.
	    if (type == TIn && l->seen[2] != TFor && l->seen[2] != TCase) {
		type = TWord;
	    }
	    // Update the current seen token type.
	    l->seen[0] = type;
	    t->type = type;
	    return t;
	}
    }
}


// lex_word scans any TWord.
static Token *lex_word(void)
{
    for (;;) {
	switch (peek()) {

	default:
	    next();
	    break;

	    // Stop scanning when any of these delimiters is found.
	case ' ':
	case '\t':
	case '\v':
	case '\f':
	case '\r':
	case '\n':
	case '\0':
	case '<':
	case '>':
	    return emit(TWord);
	}
    }
}

// lex_and scans: TAnd  TAndIf.
//                '&'   '&&'
static Token *lex_and(void)
{
    // &&
    if (peek() == '&') {
	next();
	return emit(TAndIf);
    }
    // &
    return emit(TAnd);
}

// lex_or scans: TOr  TOrIf.
//               '|'  '||'
static Token *lex_or(void)
{
    // ||
    if (peek() == '|') {
	next();
	return emit(TOrIf);
    }
    // |
    return emit(TOr);
}

// lex_semi scans: TSemi  TDSemi.
//                 ';'    ';;'
static Token *lex_semi(void)
{
    // ;;
    if (peek() == ';') {
	next();
	return emit(TDSemi);
    }
    // ;
    return emit(TSemi);
}

// lex_number scans an integer positive number that could be Tword or
// TIONumber.
static Token *lex_number(void)
{
    Token *t;
    for (;;) {
	switch (peek()) {

	default:

	    t = emit(TWord);

	    // At this point Lex->stt and Lex->pos are pointing at the start and
	    // at the end of the current positive integer in the buf line:
	    //
	    // Lex->buf = [ x | x | x | 1 | 2 | x | x | \0 ]
	    //                            ^       ^
	    //                            |       |
	    //                           stt     pos
	    //
	    // But we still don't know what type of tokin it is. It could be TWord or
	    // TIONumber. Consequently, we don't return inmediatly the token; the analysis
	    // continues.
	    goto Loop;

	case '0':
	case '1':
	case '2':
	case '3':
	case '4':
	case '5':
	case '6':
	case '7':
	case '8':
	case '9':
	    next();
	    break;
	}
    }
  Loop:

    // Consume all spaces.
    lex_space();

    char c = peek();

    // If the next character in buf is  a '<' or '>', the next token
    // could be: "<", ">", "<<", ">>", "<&", ">&", "<>", "<<-", or ">|".
    // Hence, the current token is a TIONumber.
    if (c == '<' || c == '>') {
	t->type = TIONumber;
    }

    return t;
}

// lex_less scans:  TLess  TDLess  TLessAnd  TDLessDash  TLessGreat.
//                  '<'    '<<'    '<&'      '<<-'       '<>'
static Token *lex_less(void)
{
    switch (peek()) {

	// <
    default:
	return emit(TLess);

	// << or <<-
    case '<':
	next();
	// <<-
	if (peek() == '-') {
	    next();
	    return emit(TDLessDash);
	}
	// <<
	return emit(TDLess);

	// <&
    case '&':
	next();
	return emit(TLessAnd);

	// <>
    case '>':
	next();
	return emit(TLessGreat);
    }
}

// lex_less scans:  TGreat  TDGreat  TGreatAnd  TLobber.
//                  '>'     '>>'     '>&'       '>|'
static Token *lex_great(void)
{
    switch (peek()) {

	// >
    default:
	return emit(TGreat);

	// >>
    case '>':
	next();
	return emit(TDGreat);

	// >&
    case '&':
	next();
	return emit(TGreatAnd);

	// >|
    case '|':
	next();
	return emit(TLobber);
    }
}

// lex_space consumes the space characters.
static void lex_space(void)
{
    for (;;) {
	switch (peek()) {

	default:
	    ignore();
	    return;

	    // These are space characters.
	case ' ':
	case '\t':
	case '\v':
	case '\f':
	case '\r':
	case '\n':
	    next();
	    break;
	}
    }
}

File: test/lex.lua
local ffi = require('ffi')
local lex = ffi.load('test/lex.so')

ffi.cdef [[

typedef struct __sFILE FILE;
typedef struct __sLex Lex;

typedef enum {
    TEOF,       // End of file
    TWord,      // any
    TIONumber,  // Integer positive number delimited by '<' or '>'
    TNewLine,   // \n
    TAnd,       // &
    TOr,        // |
    TSemi,      // ;
    TAndIf,     // &&
    TOrIf,      // ||
    TDSemi,     // ;;
    TLess,      // <
    TGreat,     // >
    TDLess,     // <<
    TDGreat,    // >>
    TLessAnd,   // <&
    TGreatAnd,  // >&
    TLessGreat, // <>
    TDLessDash, // <<-
    TLobber,    // >|
    TIf,        // if
    TThen,      // then
    TElse,      // else
    TElif,      // elif
    TFi,        // fi
    TDo,        // do
    TDone,      // done
    TCase,      // case
    TEsac,      // esac
    TWhile,     // while
    TUntil,     // until
    TFor,       // for
    TLBrace,    // {
    TRBrace,    // }
    TBang,      // !
    TIn,        // in
} TokenType;

typedef struct __sToken {
    char * text;
    TokenType   type;
} Token;

Lex * lex_make(void);
void lex_readfrom(const char *);
Token *lex_next(void);

]]

TokenType = {
	TEOF = 0,
	TWord = 1,
	TIONumber = 2,
	TNewLine = 3,
	TAnd = 4,
	TOr = 5,
	TSemi = 6,
	TAndIf = 7,
	TOrIf = 8,
	TDSemi = 9,
	TLess = 10,
	TGreat = 11,
	TDLess = 12,
	TDGreat = 13,
	TLessAnd = 14,
	TGreatAnd = 15,
	TLessGreat = 16,
	TDLessDash = 17,
	TLobber = 18,
	TIf = 19,
	TThen = 20,
	TElse = 21,
	TElif = 22,
	TFi = 23,
	TDo = 24,
	TDone = 25,
	TCase = 26,
	TEsac = 27,
	TWhile = 28,
	TUntil = 29,
	TFor = 30,
	TLBrace = 31,
	TRBrace = 32,
	TBang = 33,
	TIn = 34,
}

local tests = {
	{input = "0", tokens = {{type = TokenType.TWord, text = "0"}}},
	{input = "&", tokens = {{type = TokenType.TAnd, text = "&"}}},
	{input = "|", tokens = {{type = TokenType.TOr, text = "|"}}},
	{input = ";", tokens = {{type = TokenType.TSemi, text = ";"}}},
	{input = "<", tokens = {{type = TokenType.TLess, text = "<"}}},
	{input = ">", tokens = {{type = TokenType.TGreat, text = ">"}}},
	{input = "&&", tokens = {{type = TokenType.TAndIf, text = "&&"}}},
	{input = "||", tokens = {{type = TokenType.TOrIf, text = "||"}}},
	{input = ";;", tokens = {{type = TokenType.TDSemi, text = ";;"}}},
	{input = "<<", tokens = {{type = TokenType.TDLess, text = "<<"}}},
	{input = ">>", tokens = {{type = TokenType.TDGreat, text = ">>"}}},
	{input = "<&", tokens = {{type = TokenType.TLessAnd, text = "<&"}}},
	{input = ">&", tokens = {{type = TokenType.TGreatAnd, text = ">&"}}},
	{input = "<>", tokens = {{type = TokenType.TLessGreat, text = "<>"}}},
	{input = "<<-", tokens = {{type = TokenType.TDLessDash, text = "<<-"}}},
	{input = ">|", tokens = {{type = TokenType.TLobber, text = ">|"}}},
	{input = "if", tokens = {{type = TokenType.TIf, text = "if"}}},
	{input = "then", tokens = {{type = TokenType.TThen, text = "then"}}},
	{input = "else", tokens = {{type = TokenType.TElse, text = "else"}}},
	{input = "elif", tokens = {{type = TokenType.TElif, text = "elif"}}},
	{input = "fi", tokens = {{type = TokenType.TFi, text = "fi"}}},
	{input = "do", tokens = {{type = TokenType.TDo, text = "do"}}},
	{input = "done", tokens = {{type = TokenType.TDone, text = "done"}}},
	{input = "case", tokens = {{type = TokenType.TCase, text = "case"}}},
	{input = "esac", tokens = {{type = TokenType.TEsac, text = "esac"}}},
	{input = "until", tokens = {{type = TokenType.TUntil, text = "until"}}},
	{input = "while", tokens = {{type = TokenType.TWhile, text = "while"}}},
	{input = "{", tokens = {{type = TokenType.TLBrace, text = "{"}}},
	{input = "}", tokens = {{type = TokenType.TRBrace, text = "}"}}},
	{input = "!", tokens = {{type = TokenType.TBang, text = "!"}}},
	{input = "in", tokens = {{type = TokenType.TWord, text = "in"}}},
	{
		input = [[
		for word in word word word
		do
			word
			word
		done

		case word in
			;;
			;;
		esac

		while word in
		do
			word
			word
		done

		until word in
		do
			word
			word
		done

		if word in
		then
			word
			word
		elif
			word
			word
		else
			word
			word
		fi
		]], 
		tokens = {
			{type = TokenType.TFor, text = "for"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TIn, text = "in"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TDo, text = "do"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TDone, text = "done"},
			{type = TokenType.TCase, text = "case"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TIn, text = "in"},
			{type = TokenType.TDSemi, text = ";;"},
			{type = TokenType.TDSemi, text = ";;"},
			{type = TokenType.TEsac, text = "esac"},
			{type = TokenType.TWhile, text = "while"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "in"},
			{type = TokenType.TDo, text = "do"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TDone, text = "done"},
			{type = TokenType.TUntil, text = "until"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "in"},
			{type = TokenType.TDo, text = "do"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TDone, text = "done"},
			{type = TokenType.TIf, text = "if"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "in"},
			{type = TokenType.TThen, text = "then"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TElif, text = "elif"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TElse, text = "else"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TFi, text = "fi"},
		}},
	{
		input = [[
		word word<word
		word word>word
		word word<<word
		word word>>word
		word word<&word
		word word>&word
		word word<>word
		word word<<-word
		word word>|word
		word 1<2
		word 1>2
		word 1<<2
		word 1>>2
		word 1<&2
		word 1>&2
		word 1<>2
		word 1<<-2
		word 1>|2
		]],
		tokens = {
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TLess, text = "<"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TGreat, text = ">"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TDLess, text = "<<"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TDGreat, text = ">>"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TLessAnd, text = "<&"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TGreatAnd, text = ">&"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TLessGreat, text = "<>"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TDLessDash, text = "<<-"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TLobber, text = ">|"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TIONumber, text = "1"},
			{type = TokenType.TLess, text = "<"},
			{type = TokenType.TWord, text = "2"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TIONumber, text = "1"},
			{type = TokenType.TGreat, text = ">"},
			{type = TokenType.TWord, text = "2"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TIONumber, text = "1"},
			{type = TokenType.TDLess, text = "<<"},
			{type = TokenType.TWord, text = "2"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TIONumber, text = "1"},
			{type = TokenType.TDGreat, text = ">>"},
			{type = TokenType.TWord, text = "2"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TIONumber, text = "1"},
			{type = TokenType.TLessAnd, text = "<&"},
			{type = TokenType.TWord, text = "2"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TIONumber, text = "1"},
			{type = TokenType.TGreatAnd, text = ">&"},
			{type = TokenType.TWord, text = "2"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TIONumber, text = "1"},
			{type = TokenType.TLessGreat, text = "<>"},
			{type = TokenType.TWord, text = "2"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TIONumber, text = "1"},
			{type = TokenType.TDLessDash, text = "<<-"},
			{type = TokenType.TWord, text = "2"},
			{type = TokenType.TWord, text = "word"},
			{type = TokenType.TIONumber, text = "1"},
			{type = TokenType.TLobber, text = ">|"},
			{type = TokenType.TWord, text = "2"},
		}
	},
	{
		input = [[
		&&&& 
		|||| 
		;;;; 
		<<<< 
		>>>> 
		<&<&
		>&>&
		<<-<<-
		>|>|
		<><>
		]],
		tokens = {
			{type = TokenType.TAndIf, text = "&&"},
			{type = TokenType.TAndIf, text = "&&"},
			{type = TokenType.TOrIf, text = "||"},
			{type = TokenType.TOrIf, text = "||"},
			{type = TokenType.TDSemi, text = ";;"},
			{type = TokenType.TDSemi, text = ";;"},
			{type = TokenType.TDLess, text = "<<"},
			{type = TokenType.TDLess, text = "<<"},
			{type = TokenType.TDGreat, text = ">>"},
			{type = TokenType.TDGreat, text = ">>"},
			{type = TokenType.TLessAnd, text = "<&"},
			{type = TokenType.TLessAnd, text = "<&"},
			{type = TokenType.TGreatAnd, text = ">&"},
			{type = TokenType.TGreatAnd, text = ">&"},
			{type = TokenType.TDLessDash, text = "<<-"},
			{type = TokenType.TDLessDash, text = "<<-"},
			{type = TokenType.TLobber, text = ">|"},
			{type = TokenType.TLobber, text = ">|"},
			{type = TokenType.TLessGreat, text = "<>"},
			{type = TokenType.TLessGreat, text = "<>"},
		}
	}
}

print '\tlexer test:'
lex.lex_make()
for k, tt in pairs(tests) do
	lex.lex_readfrom(tt.input)

	for _, want in pairs(tt.tokens) do
		local got = lex.lex_next()

		if got.type ~= want.type or ffi.string(got.text) ~= want.text then
			print(string.format("\ttoken.text test at k=%d: got=%s, \z
				want=%s", k, ffi.string(got.text), want.text))
			print(string.format("\ttoken.type test at k=%d: got=%s, \z
				want=%s", k, got.type, want.type))
		end
	end
end
